File: bug73_late_variable_initialization/bug73_late_variable_initialization/main1.c
// Final Debugged C Code
#include <stdio.h>
#include <stdint.h>

uint64_t ror(uint64_t x, int r) {
    return (x >> r) | (x << (64 - r));
}

int main() {
    uint64_t CONST = 0x9E3779B97F4A7C15ULL;
    uint64_t mask = 0xFFFFFFFFFFFFFFFFULL;

    uint64_t acc = 0;   // must be before loop
    int step = 0;

    int data[20];
    for(int i = 0; i < 20; i++)
        data[i] = i;

    for(int i = 0; i < 20; i += 2) {
        uint64_t val = ((uint64_t)data[i] * CONST) & mask;
        acc ^= val;

        int r = (step % 9) + 1;
        acc = ror(acc, r) & mask;

        step++;
    }

    printf("0x%llx\n", acc); // string output
    return 0;
}

File: bug73_late_variable_initialization/bug73_late_variable_initialization/main1.py
# Debugged Code - Fixed late variable initialization
# The bug: accumulator was being initialized inside the loop
# Fix: initialize acc BEFORE the loop

def ror(x, r):
    # 64-bit rotate right
    return ((x >> r) | (x << (64 - r))) & 0xFFFFFFFFFFFFFFFF

def run():
    data = list(range(20))
    mask = 0xFFFFFFFFFFFFFFFF
    CONST = 0x9E3779B97F4A7C15

    acc = 0          # FIX: must be initialized before loop
    step = 0         # iteration counter for rotation

    for i in range(0, len(data), 2):
        # Multiply and keep within 64-bit
        val = (data[i] * CONST) & mask

        # XOR into accumulator
        acc ^= val

        # Rotation amount depends on iteration count, not index
        r = (step % 9) + 1

        # Rotate right 64-bit
        acc = ror(acc, r)

        step += 1

    # Output must be string (hex string)
    result = hex(acc)
    print(result)

run()